### PR TITLE
Log correlation for JBoss LogManager

### PR DESCRIPTION
**Pull request by tobiasstadler**

### CODE REVIEW

File: apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    protected void addToMdc(String key, String value) {
+        cachedRecord.get().putMdc(key, value);
+    }
+
+    @Override
+    protected void removeFromMdc(String key) {
+    }
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: Why is that necessary? Can't we directly update `org.jboss.logmanager.MDC`?

**Review by eyalkoren — CHANGES_REQUESTED** [review]
Thanks for this addition!!

I wouldn't add it this way though. I would like to have only one JBoss logging module nested under the general logging plugin. I believe that the issue here is that you need to use a different (older) dependency in order to test that. What I would do then is just add the instrumentation to the existing `apm-jboss-plugin` module and add the test, but not run it directly, but as a `TestClassWithDependencyRunner`, similarly to how [log4j 2.6 is tested](https://github.com/elastic/apm-agent-java/pull/2428/files#diff-93bf86ceebb2d3fff7cf7b7e8388cfbbd5acabebd1fc1d3d53073151c9009078).

Another alternative is to add submudules under a common JBoss logging module.

However you decide to do that, these would be the requirements:
- Have one JBoss logging module under `apm-logging-plugin`
- Make this addition as minimal as possible, including sharing as much code as possible between the two tests. They are pretty much identical, apart from how the way you get the log.

In addition, can't we share the instrumentation helper code, only updating `org.jboss.logmanager.MDC` instead of `org.jboss.logging.MDC`? If so, this logic should be shared as well.

**Comment by eyalkoren** [comment]
@tobiasstadler needless to say, if you prefer I'll take it from here- just let me know.
Your contribution is already very helpful!

**Comment by tobiasstadler** [comment]
@eyalkoren Sorry for the long delay. JBoss Logging and JBoss LogManager are two distinct things. JBoss Logging is mostly a logging facade similar to slf4j, while JBoss LogManager is an actual logging implementation which can act as a backend for JBoss Logging. But there are also adapters for e.g log4j1, log4j2 or slf4j. Maybe we should remove Instrumentation for JBoss logging and only keep one for JBoss LogManager?

File: apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java (hunk 1 above)
Comment (tobiasstadler) [reply apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: I wanted do that, but some adapters, e.g. the Log4J2 one, do call `ExtLogRecord#setMDC` when creating the `ExtLogRecord`. Since`ExtLogRecord` only "copies" the MDC once, that approach doesn’t work unfortunately.

**Comment by eyalkoren** [comment]
> Sorry for the long delay.

Extremely long - one day! I started to get worried 😜 

> JBoss Logging is mostly a logging facade similar to slf4j, while JBoss LogManager is an actual logging implementation which can act as a backend for JBoss Logging.

Right, this use case should be covered through the instrumentation of [`org.jboss.logging.JBossLogManagerLogger`](https://github.com/elastic/apm-agent-java/pull/2428/files#diff-1e60adf4201bda8a34d45d470622da8a144f6c1f53010ad9b92d779f244b0c22R56). It's only the standalone `jboss-logmanager` usage that was left unsupported.

> Maybe we should remove Instrumentation for JBoss logging and only keep one for JBoss LogManager?

Currently, this is our way to support JBoss-logging+JUL, which is the default `jboss-logging` setup. We may consider that when we add [JUL log correlation](https://github.com/elastic/apm-agent-java/issues/2492).

> JBoss Logging and JBoss LogManager are two distinct things. 

OK, then it makes sense to be in a separate module, but please nest both modules under one JBoss logging module for now. Once we get to the JUL correlation task we will check whether we can remove the current instrumentation and if that would be the case, we'll eliminate the redundant plugin level from the module hierarchy.

File: apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: I see. Then in that case, since we do not need to remove the IDs from the MDC, you can avoid caching the event as `ThreadLocal` - instead, you can change the `JBossLogManagerCorrelationHelper#cachedRecord` field to be an `ExtLogRecord` hard reference field and `JBossLogManagerCorrelationHelper#beforeLoggingEvent()` can be implemented as follows:
```
    public boolean beforeLoggingEvent(ExtLogRecord record) {
        try {
            cachedRecord = record;
            return super.beforeLoggingEvent();
        } finally {
            cachedRecord = null;
        }
    }
```
That is safe enough and it will allow removing the `afterLoggingEvent` override.

Comment (tobiasstadler) [reply apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: I don't think your approach will work when multiple threads log at the same time. As far as I can see `*CorrelationHelper` are shared by all threads so one thread might overwrite the cached record of another thread leading to incorrect correlation ids.

**Comment by tobiasstadler** [comment]
>> JBoss Logging and JBoss LogManager are two distinct things.

> OK, then it makes sense to be in a separate module, but please nest both modules under one JBoss logging module for now. Once we get to the JUL correlation task we will check whether we can remove the current instrumentation and if that would be the case, we'll eliminate the redundant plugin level from the module hierarchy.

Ok, so both Instrumentations should be in `apm-jboss-logging-plugin`? Should they also share the same instrumentation groups?

What do you think about dropping support for `org.jboss.logging.JBossLogManagerLogger` in `JBossLoggingCorrelationInstrumentation`, since this is handled in `JBossLogManagerCorrelationInstrumentation`?

**Comment by eyalkoren** [comment]
> Ok, so both Instrumentations should be in apm-jboss-logging-plugin?

Yes

> Should they also share the same instrumentation groups? 

Yes, it would be good enough, even though theoretically they can be used together, but if the need arises, we can separate them later

File: apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: Yes, you are absolutely right, it must be `ThreadLocal`, I don't know what I was thinking.
It works more elegantly with the other helpers because the event we instrumented allows direct update for the MDC, which implements the thread locality.
OK, then let's keep it this way.

Comment (tobiasstadler) [reply apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: Yes, it is a bit unfortunate, that we cannot use the MDC here.

Comment (eyalkoren) [reply apm-agent-plugins/apm-logging-plugin/apm-jboss-logmanager-plugin/src/main/java/co/elastic/apm/agent/jbosslogmanager/correlation/JBossLogManagerCorrelationHelper.java]: We could also avoid using the abstract helper or change the API to allow passing along an Object (the log event in this case), but your `TheadLocal` approach seems to be the best.
Thanks!

**Review by eyalkoren — APPROVED** [review]
Even better - together at the same module. 
Thanks!
